main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -34,19 +35,224 @@ import ca.uwaterloo.flix.util.Validation._
 object Inliner extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
 
   /**
-    * The maximum score of a function for it to be eligible for inlining.
+    * Performs inlining on the given AST `root`.
     */
-  val MaxScore = 25
+  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+    implicit val genSym: GenSym = flix.genSym
+
+    val t = System.nanoTime()
+
+    val definitions = root.definitions
+
+    /**
+      * Computes the score of each function definition
+      */
+    val scores = definitions.foldLeft(Map.empty[Symbol.DefnSym, Int]) {
+      case (macc, (sym, constant)) =>
+        macc + (sym -> Score.score(constant))
+    }
+
+    /**
+      * Transforms expressions inside definitions into optimized definitions.
+      */
+    val inlinedDefinitions = definitions.foldLeft(Map.empty[Symbol.DefnSym, Constant]) {
+      case (macc, (sym,constant)) =>
+        val newExp = inline(definitions, scores, constant.exp)
+      macc + (sym -> constant.copy(exp = newExp))
+    }
+
+    val e = System.nanoTime() - t
+    root.copy(definitions = inlinedDefinitions, time = root.time.copy(inliner = e)).toSuccess
+  }
 
   /**
-    * Performs inlining on the given AST `root`.
+    * Takes in a expression that has already been alpha-renamed and a list of args and let binds the arguments
     */
-  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+  def letBindArgs(exp: Expression, args: List[(Symbol.VarSym,Expression)]): Expression = {
+    args match {
+      case (sym,exp1) :: xs =>
+        val exp2 = letBindArgs(exp, xs)
+        Expression.Let(sym, exp1, exp2, exp2.tpe, exp1.loc)
+      case Nil =>
+        exp
+    }
+  }
 
-    // TODO: Implement.
+  /**
+    * Traverses through `exp` and performs inlining using
+    * definitions and sizes from `scores`
+    */
+  def inline(definitions: Map[Symbol.DefnSym, Constant], scores: Map[Symbol.DefnSym, Int], exp: Expression)(implicit genSym: GenSym): Expression =
+  {
+    def visit = inline(definitions, scores, _: Expression)
```

Comment:
Is this a curried version of `inline` with definitions and scores bound/captured? In these cases I tend to write an inner visitor instead.